src/java.base/share/classes/java/util/UUID.java

```diff
@@ -104,17 +104,61 @@
 
     // Constructors and Factories
 
-    /*
-     * Private constructor which uses a byte array to construct the new UUID.
+    /**
+     * Constructs new {@code UUID} from the given {@code bytes}.
+     * <p>
+     * The {@link #version()} is set to the given {@code version} and the
+     * {@link #variant()} is <strong>always</strong> set to RFC&nbsp;4122.
+     *
+     * @param bytes   to construct the {@code UUID} from.
+     * @param version to set for this {@code UUID}.
+     * @throws AssertionError if {@code bytes} is not of length 16.
+     * @since 16
+     */
+    private UUID(final byte[] bytes, final int version) {
+        assert bytes.length == 16 : "UUID bytes must be exactly of length 16, got: " + bytes.length;
+
+        long msb = (long) (bytes[0]  & 0xFF) << 56;
+        msb     |= (long) (bytes[1]  & 0xFF) << 48;
+        msb     |= (long) (bytes[2]  & 0xFF) << 40;
+        msb     |= (long) (bytes[3]  & 0xFF) << 32;
+        msb     |= (long) (bytes[4]  & 0xFF) << 24;
+        msb     |= (long) (bytes[5]  & 0xFF) << 16;
+        msb     |= (long) (bytes[6]  & 0x0F | version) << 8;
+        msb     |=         bytes[7]  & 0xFF;
+        this.mostSigBits = msb;
+
+        long lsb = (long) (bytes[8]  & 0x3F | 0x80) << 56;
+        lsb     |= (long) (bytes[9]  & 0xFF) << 48;
+        lsb     |= (long) (bytes[10] & 0xFF) << 40;
+        lsb     |= (long) (bytes[11] & 0xFF) << 32;
+        lsb     |= (long) (bytes[12] & 0xFF) << 24;
+        lsb     |= (long) (bytes[13] & 0xFF) << 16;
+        lsb     |= (long) (bytes[14] & 0xFF) << 8;
+        lsb     |=         bytes[15] & 0xFF;
+        this.leastSigBits = lsb;
+    }
+
+    /**
+     * Constructs a new {@code UUID} from the given {@code bytes}.
+     * <p>
+     * It is impossible to verify if the given data is a <em>valid</em> UUID
+     * according to RFC&nbsp;4122 or just 16 random bytes.
+     *
+     * @param bytes to construct the {@code UUID} from.
+     * @throws IllegalArgumentException if {@code bytes} is not of length 16.
+     * @since 16
      */
```

Comment:
I think you need to agreement first on whether to add this API. Maybe there is a case for a new factory method rather than a constructor? Are there any concerns with creating UUIDs with random bytes.

If it goes ahead then I assume the the javadoc will need a bit of work to specify how the byte[] is mapped to the 128-bit value.

You might need to tweak a few other things like the javadoc styles and coding style to get it consistent with the existing code. That's for later.

Reply:
Hey @AlanBateman 😊 I figured that most of the PRs I just created for UUID will require a CSR but I what couldn't figure out is/was how _I_ can create a CSR.

1. Constructor or factory, I would be fine with both and happily change the impl.
1. We de-facto already allow the construction from random data through the `UUID(long, long)` constructor.
1. No problem either.
1. I had a look at the whole JDK and various documents ([super, super, like 1999 old documents](https://www.oracle.com/java/technologies/javase/codeconventions-introduction.html)) and loaded the Oracle code style in IntelliJ. However, all I get is inconsistent. The existing files in the JDK are inconsistent. The UUID file is inconsistent in itself. The IntelliJ Oracle style is different too… I'm happy to change any line but would prefer it if there is some guidance with what is expected because I cannot figure it out on my own. 😝

Reply:
1. Changed it to static factory `valueOf`
2. —
3. I extended the documentation.
4. I tried to match the existing formatting as good as possible.